geth/node/manager.go

```diff
@@ -69,14 +72,21 @@ func (m *NodeManager) startNode(config *params.NodeConfig) (<-chan struct{}, err
 
 	m.initLog(config)
 
-	ethNode, err := MakeNode(config)
+	deliveryManager := new(notifications.DeliveryService)
+
+	ethNode, err := MakeNode(config, deliveryManager)
 	if err != nil {
 		return nil, err
 	}
 
+	m.deliveryService = deliveryManager
 	m.nodeStarted = make(chan struct{}, 1)
 
+	// Subscribe for message delivery status and log out with special key.
+	messageStateLoggingID := deliveryManager.Subscribe(logMessageStat)
+
 	go func() {
+		defer deliveryManager.Unsubscribe(messageStateLoggingID)
```

Comment:
It is calling after defer HaltOnPanic() because defers are stacking. But HaltOnPanic execute Fatalf with os.exit(1). I think deliveryManager.Unsubscribe will not execute.

Reply:
Nice catch.